### PR TITLE
improvement(parallel-operations): grow-shrink cluster

### PR DESCRIPTION
**Pull request by soyacz**
Adding and decommissioning nodes should be run in parallel as this is currently Scylla's default behavior.

Refactored code to support it. Can be disabled if
`parallel_node_operations` param is set to false.

### Testing

- [x] - provision tests
- [x] - [add 3 nodes in parallel, single AZ](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/longevity-5gb-1h-GrowShrinkClusterNemesis-aws-test/23/)
- [x] - [add 3 nodes sequentially, single az](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/longevity-5gb-1h-GrowShrinkClusterNemesis-aws-test/22/)
- [x] - [add 3 nodes in parallel, 3 az's](https://argus.scylladb.com/test/a2af091e-3cbd-4068-b825-35d511f15d17/runs?additionalRuns[]=3c8be89c-3ca1-4065-958c-c4ef0cf3ae05) - works
- [x] - [grow-shrink on gce](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/nemesis/job/longevity-5gb-1h-GrowShrinkClusterNemesis-gce-test/2/)
- [x] - [terminate and replace node](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/nemesis/job/longevity-5gb-1h-NodeTerminateAndReplace-aws-test/2/)
- [x] - [grow-shrink perf test - parallel](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/scylla-master-perf-regression-latency-650gb-grow-shrink/38/)
- [x] - [grow-shrink perf test - sequential](https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/lukasz/job/scylla-master-perf-regression-latency-650gb-grow-shrink/40/) - failed due reasons fixed by https://github.com/scylladb/scylla-cluster-tests/pull/7761 I consider it working

### PR pre-checks (self review)

- [x] I added the relevant `backport` labels
- [ ] I didn't leave commented-out/debugging code

### Reminders

- Add New configuration option and document them (in `sdcm/sct_config.py`)
- Add unit tests to cover my changes (under `unit-test/` folder)
- Update the Readme/doc folder relevant to this change (if needed)

### CODE REVIEW

**Comment by soyacz** [comment]
There's one remark about this feature: `latency_calculator_decorator` which showed 3 cycles when adding 3 nodes, now will only show one cycle. I don't know yet how to fix it - before that, I need to know if we really need it. @roydahan cc.

**Comment by fruch** [comment]
Since we were having issues that latency was bad during specific node addition 

It might be useful to keep, 

But it complicates the code leaving a whole chain of functions duplicated.

Maybe having a dedicated specific nemesis is the better option, especially when we need to support force-gossip-topology at the same time ?

**Comment by soyacz** [comment]
> Since we were having issues that latency was bad during specific node addition
> 
> It might be useful to keep,
> 
> But it complicates the code leaving a whole chain of functions duplicated.
> 
> Maybe having a dedicated specific nemesis is the better option, especially when we need to support force-gossip-topology at the same time ?

This affects only perf test - which can stay on branch perf-v14 for force-gossip-topology.
I just have got an idea how to modify code to keep support one-by-one addition/decommission. I'll update the code (or write a comment if my idea fails).

**Comment by soyacz** [comment]
> This affects only perf test - which can stay on branch perf-v14 for force-gossip-topology. I just have got an idea how to modify code to keep support one-by-one addition/decommission. I'll update the code (or write a comment if my idea fails).

ok, I think I managed to update grow-shrink nemesis to add nodes one by one, so latency wrapper should work as expected when parallel operations are off.

File: sdcm/utils/raft/__init__.py
Comment (fruch) [inline sdcm/utils/raft/__init__.py:338]: Not sure I really like the side effect in such a function

We should find a better place to use this function, and apply those changes to configure

Comment (soyacz) [reply sdcm/utils/raft/__init__.py:338]: yes, I also wasn't happy about it. I'll try to find a better place.

Comment (fruch) [reply sdcm/utils/raft/__init__.py:338]: Also one day we'll need to lock down the configuration and make it immutable during test run time.

We can apply the logic just before we gonna use the flag configuration

Comment (fruch) [reply sdcm/utils/raft/__init__.py:338]: Now I see it's not configuration...  but still confusing to some degree, cause there is such configuration

Comment (soyacz) [reply sdcm/utils/raft/__init__.py]: I removed it. Actually we need to disable `parallel_node_operations` when we force gossip - when this config param will be introduced it should be taken into account. Otherwise, version&config base should be enough for now.

File: sdcm/cluster_aws.py
Comment (vponomaryov) [inline sdcm/cluster_aws.py:403]: This condition always will return `False`, because the `rack_count` will always be a `list`.
So, it is bug that should be fixed.

**Review by vponomaryov — CHANGES_REQUESTED** [review]

File: sdcm/cluster_aws.py
Comment (soyacz) [reply sdcm/cluster_aws.py:403]: `rack_distribution` is a list which we iterate through with enumerate

File: configurations/nemesis/GrowShrinkClusterNemesis.yaml
Comment (soyacz) [reply configurations/nemesis/GrowShrinkClusterNemesis.yaml:8]: yes, here are some debug leftovers, will be removed before merge.

File: sdcm/cluster_aws.py
Comment (vponomaryov) [reply sdcm/cluster_aws.py:403]: Ah, right. Nevermind.

**Comment by vponomaryov** [comment]
LGTM, if it works.

@soyacz are the test results from the current PR state?

**Review by vponomaryov — COMMENTED** [review]
LGTM, if it works and debug part gets removed.

@soyacz are the test results from the current PR state?

**Comment by soyacz** [comment]
> LGTM, if it works and debug part gets removed.

debug part removed.

> 
> @soyacz are the test results from the current PR state?

Not much added/removed, so can say yes.

**Review by vponomaryov — APPROVED** [review]
LGTM

**Review by fruch — APPROVED** [review]

**Pull request merged**
